dev/tests/static/testsuite/Magento/Test/Integrity/Dependency/DependencyProvider.php

```diff
@@ -0,0 +1,212 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+declare(strict_types=1);
+
+namespace Magento\Test\Integrity\Dependency;
+
+use Magento\Framework\App\Utility\Files;
+use Magento\Framework\Component\ComponentRegistrar;
+
+class DependencyProvider
+{
+    /**
+     * Types of dependency between modules.
+     */
+    const TYPE_HARD = 'hard';
+
+    /**
+     * The identifier of dependency for mapping.
+     */
+    const MAP_TYPE_DECLARED = 'declared';
+
+    /**
+     * The identifier of dependency for mapping.
+     */
+    const MAP_TYPE_FOUND = 'found';
+
+    /**
+     * @var array
+     */
+    private $mapDependencies = [];
+
+    /**
+     * @var array
+     */
+    private $packageModuleMapping = [];
+
+    /**
+     * DependencyProvider constructor.
+     * @throws \Magento\Framework\Exception\LocalizedException
+     * @throws \Magento\TestFramework\Inspection\Exception
+     */
+    public function __construct()
+    {
+        $this->initDeclaredDependencies();
```

Comment:
this would make this class instantiate very slowly unless you use a Proxy class.
the other solution would be to call init before calling addDependencies or getDeclaredDependencies. basically any public method of this class

Reply:
yes, you're right, changed to the proxy.

Reply:
what I meant that instead of using a proxy ..better call that initDeclaredDependencies within addDependencies and getDeclaredDependencies and remove it from constructor.
It's even better than a proxy.

Reply:
to avoid using an init in the constructor you can move it into the method where it is needed.
e.g.
```
function getDependencies(){
    if(empty($dependencies)){
        doInit();
    }
    return $this->dependencies;
}
```